### PR TITLE
add basic support for deleting account

### LINKED ISSUE #675: Feature: Allow user to delete their Maybe account and all data
CleanShot 2024-04-25 at 11 20 40

## Feature Overview

In an effort to give users full control over their data, this issue allows them to delete their entire Maybe account and associated data.

## Requirements

_If there is a missing / incorrect requirement, please leave a comment before starting work on this._

- [ ] User can delete their user account
  - [ ] If user is `admin` and there are no other users in the family, delete the user, family, and all associated data
  - [ ] If user is `member`, only delete the user and the associated _user_ data 
  - [ ] In all other scenarios, trigger a failure
    - For example: An `admin` tries to delete their account while the family has other members.  In a scenario like this, we should error and say something like, "Admin cannot delete account while other users are present.  Please delete all members first."
- [ ] User must confirm deletion via a dialog
- [ ] After deletion, user is redirected to the sign up page with success notification
- [ ] Deletions should be atomic.  A delete action should _never_ result in invalid states or partially deleted data.
- [ ] Appropriate tests written
- [ ] i18n translations provided for `en.yml`

## Implementation Suggestions

_Below are some ideas for implementation to get you started.  Use your best judgment here—if there's a better way to do things, go for it!_

- Be sure to familiarize yourself with the `Family` and `User` models.  Currently, the app stores quite a bit of data on `Family`, hence some of the conditional deletion requirements listed above.
- Use your best judgment on this issue.  There are likely edge cases not covered here.  The overall goal at this stage is to get a _basic_ and _simple_ version of this working that can be improved upon in the future
- We _may_ want to put the deletion process in a background job, but only if it has a clear benefit

## Designs

Below are the designs you should follow while implementing this:

https://www.figma.com/file/lonJmVk3HYkwZoIO7xYP2w/Maybe-App-(Community)?type=design&node-id=3176-910&mode=design&t=4eaJRTzT6AYU5CNv-0

## Reminders

- Make sure to review our [contributing guidelines](https://github.com/maybe-finance/maybe/blob/main/CONTRIBUTING.md) before starting on an issue
- We do our best to define a clear spec for new features and fixes, but think of them as "suggestions", not "hard requirements".  We welcome ideas and suggestions!
  - If you see missing requirements to this issue, please leave a comment below explaining what is missing and why it is important.
  - If you see a requirement that you think is _incorrect_ or _not optimal_, please leave a comment explaining what you think needs to change below.

## Bounty
There is a $150 bounty on this. To participate in the bounty:

1. Comment `/attempt #675` to indicate you plan on working on this.
2. Submit a draft PR once you start and include `/claim #675` along with your general game plan for how you intent to implement this.
3. Change out of draft mode on the PR once you're ready to submit your attempt.
4. Include a short video showing functionality in your PR (drastically speeds up us approving).

In the event of multiple attempts, we generally pick the one that most closes matches the scope with the cleanest code and test coverage (when appropriate).

NOTE: Given the speed at which we're moving on the codebase, we don't assign issues or "give" issues to anyone.

### PR DESCRIPTION
**Pull request by jemiluv8**
fixes #675 
/claim #675

### DEMO 1 - admin is not allowed to delete account until there are no members
https://github.com/maybe-finance/maybe/assets/119384208/3cf8f17c-460e-49b1-a0f8-030ecc297d28

### DEMO 2 - admin account deletion after all members are deleted

https://github.com/maybe-finance/maybe/assets/119384208/86202816-9012-472e-bd7e-941cd63244b5

This draft PR demonstrates a naive implementation of user account deletion that I'll build on over the next few hours.
This implementation deletes the user and attempts to delete other related data in the same request "cycle".

The problem identified while deleting the demo account is that the delete operation can really take a long time depending on the number of accounts associated with the user's family. I have thus decided to do the cleanup in a background job.

**UI Requirements**
- [X]  User must confirm deletion via a dialog
- [X]  After deletion, user is redirected to the sign up page with success notification
- [X]  Deletions should be atomic. A delete action should never result in invalid states or partially deleted data.
- [X]  i18n translations provided for en.yml

BACKEND Requirements
- [X]  Appropriate tests written
- [X] Introduce background worker to do cleanup according to logic relating to Family, Accounts, admin and member User types
- [X] Write tests for endpoint and background worker
- [ ] Final review and testing

### CODE REVIEW

File: app/controllers/settings/profiles_controller.rb [hunk 1]
```diff
@@ -16,11 +16,32 @@ def update
       render :edit, status: :unprocessable_entity
     end
   end
+  
+  def destroy
```
Comment (jemiluv8) [inline app/controllers/settings/profiles_controller.rb:20]: the primary controller that handles deletion of the user account
It simply deletes the user account (which is the most basic requirement)
and schedules a background job to do required cleanup

File: app/javascript/controllers/application.js [hunk 2]
```diff
@@ -20,9 +20,24 @@ Turbo.setConfirmMethod((message) => {
...
 
+    if (acceptClass) {
+      document.getElementById("turbo-confirm-accept").className += acceptClass
+    }
+
     if (accept) {
       document.getElementById("turbo-confirm-accept").innerHTML = accept;
     }
+
+    if (reject) {
```
Comment (jemiluv8) [inline app/javascript/controllers/application.js:31]: Hack really to allow us too add a reject button down there

File: app/javascript/controllers/application.js [hunk 3]
```diff
@@ -20,9 +20,24 @@ Turbo.setConfirmMethod((message) => {
       document.getElementById("turbo-confirm-body").innerHTML = body;
     }
 
+    if (acceptClass) {
```
Comment (jemiluv8) [inline app/javascript/controllers/application.js:23]: Allows us to customize the accept button

File: app/jobs/delete_user_job.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        end
+      end
+    elsif user.role == "admin"
+      other_admins = User.where(family_id: user.family_id, role: "admin").where.not(id: user.id).count
+
+      if other_admins == 0
+        ActiveRecord::Base.transaction do
+          
+          if family != nil
+            family.destroy
```
Comment (jemiluv8) [inline app/jobs/delete_user_job.rb:26]: this deletes accounts and related data

File: app/jobs/delete_user_job.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+          end
+          # since other users are now invalid because of the orphaned family -> delete users as well
+          users = User.where(family_id: user.family_id)
+          users.destroy_all
```
Comment (jemiluv8) [inline app/jobs/delete_user_job.rb:30]: This is my addition thus far.
I figured since family_id is required on all user models,
any existing user with a non-existent family is "invalid"?

File: test/fixtures/users.yml [hunk 6]
```diff
@@ -4,6 +4,7 @@ family_admin:
   last_name: Dylan
   email: bob@bobdylan.com
   password_digest: <%= BCrypt::Password.create('password') %>
+  role: admin
```
Comment (jemiluv8) [inline test/fixtures/users.yml:7]: I just wonder why whoever created this test fixture never bothered to specify the admin role?

File: test/jobs/delete_user_job_test.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      DeleteUserJob.perform_later(@user_admin.attributes)
+    end
+  end
+
+  test "deleting a member with an admin will not delete the family" do
+    DeleteUserJob.perform_now(@user_member.attributes)
+    Family.find(@user_member.family_id)
+  end
+
+  test "family, related accounts and other members are deleted for the last remaining admin" do
```
Comment (jemiluv8) [inline test/jobs/delete_user_job_test.rb:27]: We could have more edge cases to test but these are it for now.

File: app/javascript/controllers/application.js [hunk 8]
```diff
@@ -20,9 +20,24 @@ Turbo.setConfirmMethod((message) => {
...
     }
+
+    if (reject) {
+      const button = document.createElement("button")
+      button.setAttribute("id", "turbo-confirm-reject")
+      button.setAttribute("class", "w-full text-red-600 rounded-xl text-center p-[10px] border mt-2")
+      button.setAttribute("value", "reject")
+      button.innerHTML = reject
+
+      document.getElementById("turbo-confirm-dialog-form").appendChild(button)
```
Comment (zachgoll) [inline app/javascript/controllers/application.js:38]: I think we could probably move this to the `_confirm_modal` partial and then toggle classes as needed here.

File: app/models/user.rb [hunk 9]
```diff
@@ -13,6 +13,14 @@ class User < ApplicationRecord
     password_salt&.last(10)
   end
 
+  def isAdmin
+    role.downcase === "admin"
+  end
+
+  def isMember
+    role.downcase === "member"
+  end
```
Comment (zachgoll) [inline app/models/user.rb:22]: These are not necessary since we've got an enum.  You should be able to already call `User.first.member?` and `User.first.admin?`.

File: app/controllers/settings/profiles_controller.rb [hunk 10]
```diff
@@ -16,11 +16,32 @@ def update
...
   def user_params
     params.require(:user).permit(:first_name, :last_name,
                                  family_attributes: [ :name, :id ])
   end
+
+  def delete_user
+    attributes = Current.user.attributes
+    Current.user.destroy!
+
+    DeleteUserJob.perform_later(attributes)
```
Comment (zachgoll) [inline app/controllers/settings/profiles_controller.rb:45]: While I think you're right that we should be doing this in a background job, I think a potentially more stable strategy would be something like this:

```rb
def mark_user_for_deletion_and_logout
  Current.user.update!(marked_for_deletion: true)
  DeleteUserJob.perform_later(Current.user)
  logout
end
```

And then on login, we could check `marked_for_deletion` and fail early if it is marked `true`.

This would allow us to work directly with the `User` record all in one spot and avoid some of the error handling here.

File: app/jobs/delete_user_job.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
```
Comment (zachgoll) [inline app/jobs/delete_user_job.rb:30]: While I think my comment above (passing in the user instance) will simplify all of this logic a bit, we should move a lot of this method into the models themselves.  As an example:

```rb
other_admins = User.where(family_id: user.family_id, role: "admin").where.not(id: user.id).count
```

This should likely be a scope or method on `Family` so that we can call:

```rb
other_admins_count = Current.family.admins.where.not(id: user.id).count
```

**Review by zachgoll — COMMENTED** [review]
I like the test cases here and overall strategy.  Just left a few suggestions on how we could simplify some of the logic.

File: app/controllers/settings/profiles_controller.rb (hunk 10 above)
Comment (zachgoll) [reply app/controllers/settings/profiles_controller.rb:45]: FYI - https://github.com/maybe-finance/maybe/issues/675#issuecomment-2081051730

File: app/jobs/delete_user_job.rb (hunk 11 above)
Comment (zachgoll) [reply app/jobs/delete_user_job.rb:30]: FYI - https://github.com/maybe-finance/maybe/issues/675#issuecomment-2081051730

File: app/models/user.rb (hunk 9 above)
Comment (jemiluv8) [reply app/models/user.rb:22]: Gotcha

File: app/javascript/controllers/application.js (hunk 8 above)
Comment (jemiluv8) [reply app/javascript/controllers/application.js:38]: I just didn't want to modify the _confirm_modal partial. I'll do that now instead. So we just toggle visibility based on args.

File: app/jobs/delete_user_job.rb (hunk 11 above)
Comment (jemiluv8) [reply app/jobs/delete_user_job.rb:30]: Gotcha.

**Comment by jemiluv8** [comment]
@zachgoll , I'll work on the changes and revert.

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -7,7 +7,7 @@ def new
   end
 
   def create
-    if user = User.authenticate_by(email: params[:email], password: params[:password])
+    if user = User.authenticate_by(email: params[:email], password: params[:password], marked_for_deletion: false)
```
Comment (jemiluv8) [inline app/controllers/sessions_controller.rb:10]: prevents users marked_for_deletion from logging in.

File: app/controllers/settings/profiles_controller.rb [hunk 13]
```diff
@@ -16,11 +16,35 @@ def update
       render :edit, status: :unprocessable_entity
     end
   end
+  
+  def destroy
+    begin
+      if Current.user.admin?
```
Comment (jemiluv8) [inline app/controllers/settings/profiles_controller.rb:22]: This is much simpler I hope.
We just refuse to delete last admin with members - otherwise all goes through

File: app/controllers/settings/profiles_controller.rb [hunk 14]
```diff
@@ -16,11 +16,35 @@ def update
...
+  end
 
   private
 
   def user_params
     params.require(:user).permit(:first_name, :last_name,
                                  family_attributes: [ :name, :id ])
   end
+
+  def mark_user_for_deletion_and_logout
```
Comment (jemiluv8) [inline app/controllers/settings/profiles_controller.rb:45]: It read so well. I couldn't help using it. Thanks for the snippet.

File: app/jobs/delete_user_job.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+class DeleteUserJob < ApplicationJob
+  queue_as :default
+
+  def perform(user)
```
Comment (jemiluv8) [inline app/jobs/delete_user_job.rb:4]: I got rid of ActiveRecord transactions here because destroys are always put within a transaction - from the docs.
And we only ever destroy here.

So far, regular users don't have anything besides access to family accounts so other cleanup might be a future thing.

Admin deletion is currently in two
- the last admin and we just get rid of the family which deletes the admin user
- not the last admin and we just delete the user only - and related data.

**Comment by jemiluv8** [comment]
@zachgoll , I've made changes based on review. Let me know how the click-testing goes.

File: app/jobs/delete_user_job.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+        else
+          user.destroy
+        end
+      end
+    end
+  rescue => e
+    Rails.logger.error "An error occurred during DeleteUserJob: #{e.message} for the user #{user.id}"
+    raise e # just propagating
+  end
+end
```
Comment (zachgoll) [inline app/jobs/delete_user_job.rb:24]: The logic here looks better, but I think we should be handling all of this within `user.rb` and `family.rb`.  Something along the lines of:

```rb
# delete_user_job.rb

def perform(user)
  user.purge_data
end
```

```rb
# user.rb

def purge_data_later
  DeleteUserJob.perform_later(self)
end

def purge_data
    if family.can_be_destroyed_by?(self)
      family.destroy! # with correct cascade behavior, this should destroy the user too
    else
      destroy!
    end
end
```

```rb
# family.rb

def can_be_destroyed_by?(user)
  users.count == 1 && user.admin?
end
```

```rb
# profiles_controller.rb
def mark_user_for_deletion_and_logout
    Current.user.update!(marked_for_deletion: true)
    Current.user.purge_data_later
    logout
end
```

File: app/views/shared/_confirm_modal.html.erb [hunk 17]
```diff
@@ -1,5 +1,5 @@
 <dialog id="turbo-confirm" class="max-w-[420px] w-full rounded-xl">
-  <form method="dialog" class="p-4">
+  <form method="dialog" class="p-4" id="turbo-confirm-dialog-form">
```
[the remark this answers is not in the document]
Comment (jemiluv8) [reply app/views/shared/_confirm_modal.html.erb:2]: It is not being used. Will clean it up.

File: app/models/family.rb [hunk 18]
```diff
@@ -4,6 +4,14 @@ class Family < ApplicationRecord
   has_many :transactions, through: :accounts
   has_many :transaction_categories, dependent: :destroy, class_name: "Transaction::Category"
 
+  def admins
+    users.where(users: { role: "admin", marked_for_deletion: false })
+  end
+
+  def members
+    users.where(users: { role: "member", marked_for_deletion: false })
+  end
```
[the remark this answers is not in the document]
Comment (jemiluv8) [reply app/models/family.rb:18]: updated the method to not user the marked_for_deletion query.
Had challenges using scope here though so maintained the method.

**Comment by zachgoll** [comment]
@jemiluv8 thanks for the updates here.  I think this PR is at a final state now, although in #698 there is a "deactivate" strategy implemented which I think offers a more concise and intuitive way to handle what we had come up here with the `marked_for_deletion` flag on the `User` model.

Since there's a lot of good work here that allowed us to see the requirements for this issue more clearly and come to a solid solution, we'll award a split bounty on this.

I'll close this PR out, but there will be a payment awarded after #698 is merged.  Thanks for all the work on this!

**Comment by jemiluv8** [comment]
That's great @zachgoll, I learned a lot and been intrigued by the deactivation strategy.
It looked solid.

**Comment by zachgoll** [comment]
@jemiluv8 just an FYI, the bounty split didn't work properly when the PR was merged, but we're working to get that resolved right now so you get the appropriate payout.
